### PR TITLE
Implement Cluster API

### PR DESCRIPTION
**Pull request by NavisJ**
# PR Submission checklist

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
|https://github.com/dell/csm/issues/888|

# Common PR Checklist:

- [x] Have you made sure that the code compiles?
- [x] Have you commented your code, particularly in hard-to-understand areas
- [ ] Did you run tests in a real Kubernetes cluster?
- [x] Have you maintained backward compatibility

## Tests

```
time="2023-07-13T13:48:16+08:00" level=debug msg="opts.Insecure : 'true'"
time="2023-07-13T13:48:16+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP REQUEST -------------------------\n    GET /platform/latest/ HTTP/1.1\n    Host: 10.225.108.6:8080\n    Authorization: root:******\n    \n"
time="2023-07-13T13:48:17+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP RESPONSE -------------------------\n    HTTP/1.1 200 Ok\n    Transfer-Encoding: chunked\n    Content-Security-Policy: default-src 'self' 'unsafe-inline' 'unsafe-eval' data:; script-src 'self' 'unsafe-eval'; style-src 'unsafe-inline' 'self';\n    Content-Type: application/json\n    Date: Thu, 13 Jul 2023 04:28:04 GMT\n    Server: Apache\n    Strict-Transport-Security: max-age=31536000;\n    X-Content-Type-Options: nosniff\n    X-Frame-Options: sameorigin\n    X-Xss-Protection: 1; mode=block\n    "
=== RUN   TestGetStatistics
time="2023-07-13T13:48:17+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP REQUEST -------------------------\n    GET /platform/3/statistics/current?keys=ifs.bytes.avail%!C(MISSING)ifs.bytes.total HTTP/1.1\n    Host: 10.225.108.6:8080\n    Authorization: root:******\n    \n"
time="2023-07-13T13:48:18+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP RESPONSE -------------------------\n    HTTP/1.1 200 Ok\n    Transfer-Encoding: chunked\n    Allow: GET, HEAD\n    Content-Security-Policy: default-src 'self' 'unsafe-inline' 'unsafe-eval' data:; script-src 'self' 'unsafe-eval'; style-src 'unsafe-inline' 'self';\n    Content-Type: application/json\n    Date: Thu, 13 Jul 2023 04:28:05 GMT\n    Server: Apache\n    Strict-Transport-Security: max-age=31536000;\n    X-Content-Type-Options: nosniff\n    X-Frame-Options: sameorigin\n    X-Xss-Protection: 1; mode=block\n    "
--- PASS: TestGetStatistics (1.02s)
=== RUN   TestGetFloatStatistics
time="2023-07-13T13:48:18+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP REQUEST -------------------------\n    GET /platform/3/statistics/current?keys=cluster.disk.bytes.in.rate%!C(MISSING)ifs.bytes.total%!C(MISSING)cluster.disk.xfers.in.rate HTTP/1.1\n    Host: 10.225.108.6:8080\n    Authorization: root:******\n    \n"
time="2023-07-13T13:48:19+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP RESPONSE -------------------------\n    HTTP/1.1 200 Ok\n    Transfer-Encoding: chunked\n    Allow: GET, HEAD\n    Content-Security-Policy: default-src 'self' 'unsafe-inline' 'unsafe-eval' data:; script-src 'self' 'unsafe-eval'; style-src 'unsafe-inline' 'self';\n    Content-Type: application/json\n    Date: Thu, 13 Jul 2023 04:28:06 GMT\n    Server: Apache\n    Strict-Transport-Security: max-age=31536000;\n    X-Content-Type-Options: nosniff\n    X-Frame-Options: sameorigin\n    X-Xss-Protection: 1; mode=block\n    "
--- PASS: TestGetFloatStatistics (1.06s)
=== RUN   TestGetLocalSerial
time="2023-07-13T13:48:19+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP REQUEST -------------------------\n    GET /platform/3/cluster/config/ HTTP/1.1\n    Host: 10.225.108.6:8080\n    Authorization: root:******\n    \n"
time="2023-07-13T13:48:20+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP RESPONSE -------------------------\n    HTTP/1.1 200 Ok\n    Transfer-Encoding: chunked\n    Allow: GET, HEAD\n    Content-Security-Policy: default-src 'self' 'unsafe-inline' 'unsafe-eval' data:; script-src 'self' 'unsafe-eval'; style-src 'unsafe-inline' 'self';\n    Content-Type: application/json\n    Date: Thu, 13 Jul 2023 04:28:07 GMT\n    Server: Apache\n    Strict-Transport-Security: max-age=31536000;\n    X-Content-Type-Options: nosniff\n    X-Frame-Options: sameorigin\n    X-Xss-Protection: 1; mode=block\n    "
SV200-004EIH-G7ON
--- PASS: TestGetLocalSerial (1.08s)
=== RUN   TestGetClusterConfig
time="2023-07-13T13:48:20+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP REQUEST -------------------------\n    GET /platform/3/cluster/config/ HTTP/1.1\n    Host: 10.225.108.6:8080\n    Authorization: root:******\n    \n"
time="2023-07-13T13:48:22+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP RESPONSE -------------------------\n    HTTP/1.1 200 Ok\n    Transfer-Encoding: chunked\n    Allow: GET, HEAD\n    Content-Security-Policy: default-src 'self' 'unsafe-inline' 'unsafe-eval' data:; script-src 'self' 'unsafe-eval'; style-src 'unsafe-inline' 'self';\n    Content-Type: application/json\n    Date: Thu, 13 Jul 2023 04:28:08 GMT\n    Server: Apache\n    Strict-Transport-Security: max-age=31536000;\n    X-Content-Type-Options: nosniff\n    X-Frame-Options: sameorigin\n    X-Xss-Protection: 1; mode=block\n    "
--- PASS: TestGetClusterConfig (1.04s)
=== RUN   TestGetClusterIdentity
time="2023-07-13T13:48:22+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP REQUEST -------------------------\n    GET /platform/3/cluster/identity/ HTTP/1.1\n    Host: 10.225.108.6:8080\n    Authorization: root:******\n    \n"
time="2023-07-13T13:48:23+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP RESPONSE -------------------------\n    HTTP/1.1 200 Ok\n    Transfer-Encoding: chunked\n    Allow: GET, PUT, HEAD\n    Content-Security-Policy: default-src 'self' 'unsafe-inline' 'unsafe-eval' data:; script-src 'self' 'unsafe-eval'; style-src 'unsafe-inline' 'self';\n    Content-Type: application/json\n    Date: Thu, 13 Jul 2023 04:28:09 GMT\n    Server: Apache\n    Strict-Transport-Security: max-age=31536000;\n    X-Content-Type-Options: nosniff\n    X-Frame-Options: sameorigin\n    X-Xss-Protection: 1; mode=block\n    "
--- PASS: TestGetClusterIdentity (1.04s)
=== RUN   TestGetClusterAcs
time="2023-07-13T13:48:23+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP REQUEST -------------------------\n    GET /platform/14/cluster/acs/ HTTP/1.1\n    Host: 10.225.108.6:8080\n    Authorization: root:******\n    \n"
time="2023-07-13T13:48:24+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP RESPONSE -------------------------\n    HTTP/1.1 200 Ok\n    Transfer-Encoding: chunked\n    Allow: GET, POST, HEAD\n    Content-Security-Policy: default-src 'self' 'unsafe-inline' 'unsafe-eval' data:; script-src 'self' 'unsafe-eval'; style-src 'unsafe-inline' 'self';\n    Content-Type: application/json\n    Date: Thu, 13 Jul 2023 04:28:10 GMT\n    Server: Apache\n    Strict-Transport-Security: max-age=31536000;\n    X-Content-Type-Options: nosniff\n    X-Frame-Options: sameorigin\n    X-Xss-Protection: 1; mode=block\n    "
--- PASS: TestGetClusterAcs (1.01s)
=== RUN   TestGetClusterInternalNetworks
time="2023-07-13T13:48:24+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP REQUEST -------------------------\n    GET /platform/7/cluster/internal-networks/ HTTP/1.1\n    Host: 10.225.108.6:8080\n    Authorization: root:******\n    \n"
time="2023-07-13T13:48:25+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP RESPONSE -------------------------\n    HTTP/1.1 200 Ok\n    Transfer-Encoding: chunked\n    Allow: GET, PUT, HEAD\n    Content-Security-Policy: default-src 'self' 'unsafe-inline' 'unsafe-eval' data:; script-src 'self' 'unsafe-eval'; style-src 'unsafe-inline' 'self';\n    Content-Type: application/json\n    Date: Thu, 13 Jul 2023 04:28:11 GMT\n    Server: Apache\n    Strict-Transport-Security: max-age=31536000;\n    X-Content-Type-Options: nosniff\n    X-Frame-Options: sameorigin\n    X-Xss-Protection: 1; mode=block\n    "
--- PASS: TestGetClusterInternalNetworks (1.01s)
=== RUN   TestGetClusterNodes
time="2023-07-13T13:48:25+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP REQUEST -------------------------\n    GET /platform/3/cluster/nodes/ HTTP/1.1\n    Host: 10.225.108.6:8080\n    Authorization: root:******\n    \n"
time="2023-07-13T13:48:26+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP RESPONSE -------------------------\n    HTTP/1.1 200 Ok\n    Transfer-Encoding: chunked\n    Allow: GET, HEAD\n    Content-Security-Policy: default-src 'self' 'unsafe-inline' 'unsafe-eval' data:; script-src 'self' 'unsafe-eval'; style-src 'unsafe-inline' 'self';\n    Content-Type: application/json\n    Date: Thu, 13 Jul 2023 04:28:12 GMT\n    Server: Apache\n    Strict-Transport-Security: max-age=31536000;\n    X-Content-Type-Options: nosniff\n    X-Frame-Options: sameorigin\n    X-Xss-Protection: 1; mode=block\n    "
--- PASS: TestGetClusterNodes (5.02s)
=== RUN   TestGetClusterNode
time="2023-07-13T13:48:30+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP REQUEST -------------------------\n    GET /platform/3/cluster/nodes/1/ HTTP/1.1\n    Host: 10.225.108.6:8080\n    Authorization: root:******\n    \n"
time="2023-07-13T13:48:31+08:00" level=debug msg="\n    -------------------------- GOISILON HTTP RESPONSE -------------------------\n    HTTP/1.1 200 Ok\n    Transfer-Encoding: chunked\n    Allow: GET, PUT, HEAD\n    Content-Security-Policy: default-src 'self' 'unsafe-inline' 'unsafe-eval' data:; script-src 'self' 'unsafe-eval'; style-src 'unsafe-inline' 'self';\n    Content-Type: application/json\n    Date: Thu, 13 Jul 2023 04:28:17 GMT\n    Server: Apache\n    Strict-Transport-Security: max-age=31536000;\n    X-Content-Type-Options: nosniff\n    X-Frame-Options: sameorigin\n    X-Xss-Protection: 1; mode=block\n    "
--- PASS: TestGetClusterNode (1.88s)
PASS


Process finished with the exit code 0

```

## Description of your changes:
- Add onefs version and timezone variables to cluster config API
- Add cluster identity API
- Add cluster ACS API
- Add cluster internal networks API
- Add API for query single and multiple cluster nodes

### CODE REVIEW

File: cluster_test.go [hunk 1]
```diff
@@ -49,3 +52,58 @@ func TestGetLocalSerial(t *testing.T) {
...
+	}
+	log.Debug(defaultCtx, "Get identity name of the cluster '%s'", identity.Name)
+}
+
+func TestGetClusterAcs(t *testing.T) {
+	acs, err := client.GetClusterAcs(defaultCtx)
+	if err != nil {
+		panic(err)
+	}
+	log.Debug(defaultCtx, "%+v", acs)
```
Comment (taohe1012) [inline cluster_test.go:79]: better not use log.Debug to print info, please use assert here for checking

Comment (taohe1012) [reply cluster_test.go:79]: And other tests also have similar situation, please update

Comment (NavisJ) [reply cluster_test.go]: Updated

File: cluster.go [hunk 2]
```diff
@@ -76,6 +82,51 @@ func (c *Client) GetClusterConfig(ctx context.Context) (ClusterConfig, error) {
...
+func (c *Client) GetClusterInternalNetworks(ctx context.Context) (ClusterInternalNetworks, error) {
+	clusterInternalNetworks, err := apiv7.GetIsiClusterInternalNetworks(ctx, c.API)
+	if err != nil {
+		return nil, err
+	}
+	return clusterInternalNetworks, nil
+}
+
+// GetIsiClusterNodes list the nodes on this cluster
+func (c *Client) GetIsiClusterNodes(ctx context.Context) (ClusterNodes, error) {
```
[the remark this answers is not in the document]
Comment (NavisJ) [reply cluster.go]: Updated the method name and testing result

File: cluster_test.go [hunk 3]
```diff
@@ -49,3 +52,45 @@ func TestGetLocalSerial(t *testing.T) {
...
+	assertNotNil(t, nodes)
+	assert.NotEqual(t, int(*nodes.Total), 0)
+}
+
+func TestGetIsiClusterNode(t *testing.T) {
+	var nodeID = 1
+	nodes, err := client.GetIsiClusterNode(defaultCtx, nodeID)
+	assertNoError(t, err)
+	assertNotNil(t, nodes)
+	assert.Equal(t, int(*nodes.Total), 1)
```
[the remark this answers is not in the document]
Comment (NavisJ) [reply cluster_test.go]: Changed

File: cluster_test.go [hunk 4]
```diff
@@ -49,3 +52,45 @@ func TestGetLocalSerial(t *testing.T) {
...
+	networks, err := client.GetClusterInternalNetworks(defaultCtx)
+	assertNoError(t, err)
+	assertNotNil(t, networks)
+}
+
+func TestGetIsiClusterNodes(t *testing.T) {
+	nodes, err := client.GetIsiClusterNodes(defaultCtx)
+	assertNoError(t, err)
+	assertNotNil(t, nodes)
+	assert.NotEqual(t, int(*nodes.Total), 0)
```
[the remark this answers is not in the document]
Comment (NavisJ) [reply cluster_test.go]: Changed

**Review by taohe1012 — APPROVED** [review]

**Review by shefali-malhotra — COMMENTED** [review]
Are the new test cases added to maintain the coverage?

**Comment by NavisJ** [comment]
> Are the new test cases added to maintain the coverage?

Yes. Corresponding test cases have been added for every new cluster API function.

**Review by chimanjain — APPROVED** [review]

**Review by shefali-malhotra — APPROVED** [review]

**Pull request merged**